### PR TITLE
[SYCL] Fix deadlock in ProgramManager class

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch fixes deadlock in ```sycl::detail::ProgramManager``` class caused by data race in usage of ```KernelProgramCache::MBuildCV``` condition variable.
Even if ```BuildResult->State.load()``` is atomic, it must be modified under the mutex. The deadlock happens in situation when notification ```notify_all()``` is sent while ```MBuildCV``` is in the wait expression BUT not in the waiting state. The result is - the notification is lost. After that the thread goes to the waiting state and sleeps forever.
```BuildResult``` is protected by mutex now, so the notification will be sent only when the other thread in the waiting state.

### CODE REVIEW

**Comment by dm-vodopyanov** [comment]
@kbobrovs, @rbegam, ping.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -414,7 +416,8 @@ ProgramManager::getOrCreateKernel(OSModuleHandle M, const context &Context,
                                   const program_impl *Prg) {
   if (DbgProgMgr > 0) {
     std::cerr << ">>> ProgramManager::getOrCreateKernel(" << M << ", "
-              << getRawSyclObjImpl(Context) << ", " << KernelName << ")\n";
+              << getRawSyclObjImpl(Context) << ", " << KernelName << ", "
+              << Prg->get() << ")\n";
```
Comment (s-kanaev) [inline sycl/source/detail/program_manager/program_manager.cpp:420]: Is this change needed in this patch?
I suggest moving it to distinct patch.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 2]
```diff
@@ -82,9 +75,9 @@ class KernelProgramCache {
     return {MKernelsPerProgramCache, MKernelsPerProgramCacheMutex};
   }
 
-  template <class Predicate> void waitUntilBuilt(Predicate Pred) const {
-    std::unique_lock<std::mutex> Lock(MBuildCVMutex);
-
+  template <class Predicate>
+  void waitUntilBuilt(std::unique_lock<std::mutex> &Lock,
```
Comment (s-kanaev) [inline sycl/source/detail/kernel_program_cache.hpp:79]: I suggest trying to rearrange load, store, exchange operations to fit in proper order instead of employing distinct mutex.

Implementation of this approach is to check atomic variable after `MBuildCVMutex` is locked as `lock` operation may take some time. This approach will require a distinct lambda with check operation as an argument to `waitUntilBuilt()` method.

If this approach doesn't fit, I think the distinct mutex shown in current version of the patch is as good as a solution.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 3]
```diff
@@ -33,21 +33,20 @@ class KernelProgramCache {
...
-  template<typename T> struct BuildResult {
+  template <typename T> struct BuildResult {
     std::atomic<T *> Ptr;
     std::atomic<int> State;
     BuildError Error;
+    std::mutex MBuildResultMutex;
 
-    BuildResult(T* P, int S) : Ptr{P}, State{S}, Error{"", 0} {}
+    BuildResult(T *P, int S) : Ptr{P}, State{S}, Error{"", 0} {}
   };
```
Comment (kbobrovs) [inline sycl/source/detail/kernel_program_cache.hpp:50]: Please move all the above format changes to a separate patch

File: sycl/source/detail/kernel_program_cache.hpp (hunk 2 above)
Comment (kbobrovs) [inline sycl/source/detail/kernel_program_cache.hpp:79]: I think the concurrent cache management is quite complex piece of design which is worth a section in the design doc. Can you guys take an effort to write one? That would really help to do a good review.

Comment (dm-vodopyanov) [reply sycl/source/detail/kernel_program_cache.hpp:79]: What do you mean by checking the atomic variable? (is it```load()``` or something?)
I'm not quite understand, if there is no the distinct mutex for ```BuildResult```, will this approach fix the situation when notification sends before putting another thread to waiting state?

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/program_manager/program_manager.cpp]: Ok, will add this enhancement for debugging ProgramManager as separate PR.

File: sycl/source/detail/kernel_program_cache.hpp (hunk 3 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/kernel_program_cache.hpp:52]: Done

File: sycl/source/detail/kernel_program_cache.hpp (hunk 2 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/kernel_program_cache.hpp:81]: We can prepare a doc when @s-kanaev will be available, as he developed this solution. For now, my understanding is the following. 
When we want to get some kernel or program from cache, we first lock cache and check in which build state kernel or program is right now (Done, Failed or In Progress). 
If it is already In Progress state in the cache, when some other thread already added kernel or program to cache, in this case we move current thread to waiting state. It will be woken up by notification from other thread that build was Done or build was Failed. If build was Done, we return build result, otherwise (when build was Failed), we try one more time to build it successfully.
If there was no information about the kernel or program in the cache in In Progress state, we need to simply build it, when save the result and move build state to Done, when current thread notifies other threads that build was completed and they can wake up.
Above is explanation of how ```getOrBuild(...)``` function from ```program_manager.cpp``` works, deadlock was happened in this function. Wait and notification mechanisms utilize ```std::condition variable```, build result stores to atomic variables ```BuildResult->Ptr``` and ```BuildResult->State```.

Comment (s-kanaev) [reply sycl/source/detail/kernel_program_cache.hpp:81]: > What do you mean by checking the atomic variable? (is it `load()` or something?)

Yes, it's  `load()` exactly.

> I'm not quite understand, if there is no the distinct mutex for ```BuildResult```, will this approach fix the situation when notification sends before putting another thread to waiting state?


`store()` operation for state of built result precedes call to `KernelProgramCache::notifyAllBuild()`.

As far as I understand the use case is the following:
 * Thread1 is building program P1. The program is built by this thread. Right now the thread is going to store build state.
 * Thread2 is building program P2. Right now it's executing `KernelProgramCache::waitUntilBuilt()`, it's in the call to `condition_variable::wait()`.
 * Thread3 is building program P1. Right now it's executing `KernelProgramCache::waitUntilBuilt()`, it's in the call to `condition_variable::wait()`.
 * Thread4 is building program P2. The program is built by this thread. Right now the thread is executing `KernelProgramCache::notifyAllBuild()` function and `condition_variable::notify_all()` is already executed.

The deadlock is going to appear in Thread3.

Now, Thread2 and Thread3 are waked up. Thread3 now owns `KernelProgramCache::MBuildCVMutex` mutex. Thread2 is in waiting state on the mutex now. Thread3 checks the state and sees `BS_InProgress` there, unblocks the mutex. Thread1 stores build state and calls to `condition_variable::notify_all()`. Thread2 locks the mutex and check build state and sees `BS_Done`, unblocks the mutex and returns from `condition_variable::wait()`. Thread3 enters waiting state on condition variable until either a spurious wakeup or some other thread calls `condition_variable::notify_all()/notify_one()`.

Now I see, why my suggestion won't help here.

Although, I have another suggestion. Let's make your solution complete. Currently we use a common `condition_variable` for each and every build. We can use a per-result `condition_variable`. This will eliminate wakeups of threads waiting on program P1 build when some thread signals on build of P2. Though there may be a restriction on number of condition variables. This restriction will originate from OS. So we should check if there is no limit on the number of condition variables (i.e. the only resource it employs is RAM).

Comment (s-kanaev) [reply sycl/source/detail/kernel_program_cache.hpp:81]: @kbobrovs The idea is described by @dm-vodopyanov quite correctly.

Comment (dm-vodopyanov) [reply sycl/source/detail/kernel_program_cache.hpp:81]: I'm not sure if we can use multiple ```condition_variable```s because they are not CopyConstructible, MoveConstructible, CopyAssignable or MoveAssignable.
What I did: 
1. I created two new members of ```KernelProgramCache``` class:
```c++
using ProgramBuildCVT = std::map<ProgramCacheKeyT, std::condition_variable>;
using KernelBuildCVT = std::map<RT::PiProgram, std::map<string_class, std::condition_variable>>;
...
auto& acquireProgramConditionVariables() { return MProgramBuildCVs; }
auto& acquireKernelConditionVariables() { return MKernelBuildCVs; }
...
ProgramBuildCVT MProgramBuildCVs;
KernelBuildCVT MKernelBuildCVs;
```
2. Then in passed new functor to ```getOrBuild()``` function:

* for programs:

```c++
auto CondVarsF = [](KernelProgramCache &Cache) {
  return Cache.acquireProgramConditionVariables();
};
```

* for kernels:

```c++
auto CondVarsF = [&Program](KernelProgramCache &Cache) {
  return Cache.acquireKernelConditionVariables()[Program];
};
```
3. In ```getOrBuild()``` function I inserted the new pair to MProgramBuildCVs or MKernelBuildCVs:
```c++
CondVars(KPCache).emplace(std::piecewise_construct, std::forward_as_tuple(CacheKey), std::forward_as_tuple(std::condition_variable()));
```
and got compilation problem here because copy constructor of ```std::condition_variable``` marked as deleted. It looks like we can't create a map with ```std::condition_variable``` inside.
After that I wanted to get the right condition variable like this:
```c++
std::condition_variable &BuildCV = CondVars(KPCache).at(CacheKey);
```
as pass it to ```waitUntilBuilt()``` and ```notifyAllBuild()```.

What do you think about this? I also thought about adding new field ```std::condition_variable MBuildCV;``` to ```BuildResult``` but it looks like it doesn't work because the object of ```BuildResult``` struct is unique for each thread when we enter ```getOrBuild()``` function.

Comment (s-kanaev) [reply sycl/source/detail/kernel_program_cache.hpp:81]: > What do you think about this? I also thought about adding new field ```std::condition_variable MBuildCV;``` to ```BuildResult``` but it looks like it doesn't work because the object of ```BuildResult``` struct is unique for each thread when we enter ```getOrBuild()``` function.

`BuildResult` should not be unique per thread. It should only be unique per building object. Note, that insertion of `BuildResult` object may not take place really (see the section of `getOrBuild()` where the call to `Cache.emplace()` takes place). Note also, that `BuildResult` is a pointer to value which contains in the cache (which is a map, actually). Hence, you may augment constructor of `BuildResult` to initialize a `condition_variable` member variable without changing its signature.

Comment (dm-vodopyanov) [reply sycl/source/detail/kernel_program_cache.hpp]: Ah, ok, thanks a lot! I debugged it incorrectly.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 4]
```diff
@@ -82,21 +79,19 @@ class KernelProgramCache {
...
-  template <class Predicate> void waitUntilBuilt(Predicate Pred) const {
-    std::unique_lock<std::mutex> Lock(MBuildCVMutex);
-
-    MBuildCV.wait(Lock, Pred);
+  template <class Predicate>
+  void waitUntilBuilt(std::condition_variable &CV,
+                      std::unique_lock<std::mutex> &Lock,
+                      Predicate Pred) const {
+    CV.wait(Lock, Pred);
   }
```
Comment (s-kanaev) [inline sycl/source/detail/kernel_program_cache.hpp:87]: I believe this function should have a comment, saying that `condition_variable` and `unique_lock` should be for the same `BuildResult`.

Comment (s-kanaev) [inline sycl/source/detail/kernel_program_cache.hpp:87]: Also, as far as I see you can leave the contents of this function as it is, i.e. lock the appropriate `mutex` for `BuildResult` and await on its `condition_variable`. I mean, I can't find changes in `program_manager.cpp` which require this kind of change of signature of the function.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 5]
```diff
@@ -82,21 +79,19 @@ class KernelProgramCache {
     return {MKernelsPerProgramCache, MKernelsPerProgramCacheMutex};
   }
 
-  template <class Predicate> void waitUntilBuilt(Predicate Pred) const {
-    std::unique_lock<std::mutex> Lock(MBuildCVMutex);
-
-    MBuildCV.wait(Lock, Pred);
+  template <class Predicate>
+  void waitUntilBuilt(std::condition_variable &CV,
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/source/detail/kernel_program_cache.hpp:83]: I think it hides the details of cache implementation.
Though, the methods are in a lack of documentation. I believe we'll add some documentation for the cache soon.
